gramhopper/configuration/trigger_response_parser.py

```diff
@@ -0,0 +1,121 @@
+import abc
+from typing import Dict, get_type_hints, Union, List, Any, Type
+from inspect import isclass
+from .partial_ruamel_yaml import CommentedMap, CommentedSeq
+from .common_types import TriggerResponse, GlobalsDict
+from .boolean_helper import BooleanHelper
+from ..dict_enum import DictEnum
+
+
+class TriggerResponseParser(abc.ABC):
+    """ A common base parser for trigger/response configuration parsers"""
+
+    def __init__(self, mapping_class: Type[DictEnum], element_base_class: Type[TriggerResponse]):
+        self.__mapping_class = mapping_class
+        self.__element_base_class = element_base_class
+
+    @property
+    def mapping_class(self) -> Type[DictEnum]:
+        """ Returns the mapping class (`Triggers` or `Responses`)"""
+        return self.__mapping_class
+
+    @property
+    def element_base_class(self) -> Type[TriggerResponse]:
+        """ Returns the element base class (`BaseTrigger` or `BaseResponse`)"""
+        return self.__element_base_class
+
+    def parse_atomic(self, config: Dict[str, Any]) -> TriggerResponse:
+        """
+        Parses an atomic configuration subtree as a trigger/response
+        :param config: A configuration subtree to parse
+        :return: The trigger/response object created from the given configuration subtree
+        """
+        config_copy = dict(config)
+        name = None
+        if 'name' in config_copy:
+            name = config_copy.pop('name')
+
+        element = self.mapping_class[config_copy.pop('type')]
+
+        # Some triggers (most of them) are classes and some are instances (mostly filter triggers).
+        # This allows both cases to be used.
+        if isclass(element):
+            trigger_or_response = element(**config_copy)
+            if name:
+                trigger_or_response.name = name
+            return trigger_or_response
+        return element
+
+    def parse_single(self, config: Union[CommentedMap, str], global_elements: GlobalsDict) \
+            -> TriggerResponse:
+        """
+        Recursively parses a single configuration subtree as a trigger/response
+        :param config: A configuration subtree to parse
+        :param global_elements: A dictionary with all triggers and responses configured globally
+        :return: The trigger/response object created from the given configuration subtree
+        """
+        if isinstance(config, str):
+            return BooleanHelper.parse_boolean_subrule(config, global_elements)
+
+        config_copy = dict(config)
```

Comment:
What kind of copy?

Reply:
What do you mean?

Reply:
deep or shallow? 
what is the original type of `config` , also a dict?

Reply:
Shallow.
The original type is `CommentedMap`, which inherits `dict`.